ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java

```diff
@@ -169,54 +208,118 @@ void testConnectivity() {
   }
 
   /**
-   * Get the list of blobs in the specified partition matching the specified DocumentDB query.
+   * Get the list of blobs in the specified partition that have been deleted or expired for at least the
+   * configured retention period.
    * @param partitionPath the partition to query.
-   * @param querySpec the DocumentDB query to execute.
-   * @param timer the {@link Timer} to use to record query time (excluding waiting).
-   * @return a List of {@link CloudBlobMetadata} referencing the matching blobs.
+   * @param fieldName the field name to query on. Allowed values are {@link CloudBlobMetadata#FIELD_DELETION_TIME} and
+   *                  {@link CloudBlobMetadata#FIELD_EXPIRATION_TIME}.
+   * @param retentionThreshold the latest time where blobs are considered dead if they were expired
+   *                           or deleted before that point.
+   * @param maxEntries the max number of metadata records to return.
+   * @return a List of {@link CloudBlobMetadata} referencing the dead blobs found.
+   * @throws CloudStorageException
    */
-  List<CloudBlobMetadata> queryMetadata(String partitionPath, SqlQuerySpec querySpec, Timer timer)
+  List<CloudBlobMetadata> getDeadBlobs(String partitionPath, String fieldName, long retentionThreshold, int maxEntries)
       throws DocumentClientException {
-    azureMetrics.documentQueryCount.inc();
+
+    String deadBlobsQuery;
+    if (fieldName.equals(CloudBlobMetadata.FIELD_DELETION_TIME)) {
+      deadBlobsQuery = DELETED_BLOBS_QUERY;
+    } else if (fieldName.equals(CloudBlobMetadata.FIELD_EXPIRATION_TIME)) {
+      deadBlobsQuery = EXPIRED_BLOBS_QUERY;
+    } else {
+      throw new IllegalArgumentException("Invalid field: " + fieldName);
+    }
+    SqlQuerySpec querySpec = new SqlQuerySpec(deadBlobsQuery,
+        new SqlParameterCollection(new SqlParameter(LIMIT_PARAM, maxEntries),
+            new SqlParameter(THRESHOLD_PARAM, retentionThreshold)));
+
     FeedOptions feedOptions = new FeedOptions();
+    feedOptions.setMaxItemCount(maxEntries);
+    feedOptions.setResponseContinuationTokenLimitInKb(continuationTokenLimit);
```

Comment:
What happens when the continuation token gets bigger than this limit? Does the query iterator stop returning new values until another query is made?

Reply:
No, acc to MS it just makes subsequent queries (during paging) less efficient if it can't stuff as much state in the token.  Doubt it will affect us much.